### PR TITLE
feat: Implement prometheus pull monitoring metrics data from the openGemini

### PR DESCRIPTION
**Pull request by gueFDF**
…table and write them to Prometheus

### What problem does this PR solve?
Implement openGeminicon to read monitoring indicators from the table and write them to Prometheus

### What is changed and how it works?
Implemented the /metrics interface to read metrics from db for reporting.

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

### How to use it?
Use the following example.
Add the following to the configuration file.
```toml
[monitor]
pushers = "http"
store-enabled = true
store-database = "_internal"
store-interval = "10s"
store-path = "/tmp/openGemini/metric/ts-sql/metric.data"
http-endpoint = "127.0.0.1:8086"
```
At this point, can get the metrics from an HTTP request
```shell
"http://localhost:8086/metrics"
```

### CODE REVIEW

File: lib/util/lifted/influx/httpd/handler_metrics.go [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+	"go.uber.org/zap"
+)
+
+var (
+	metricsHandler      http.Handler
+	openGeminiCollector *OpenGeminiCollector
+)
+
+func init() {
+	// 创建一个自定义的注册表
```
Comment (xiangyu5632) [inline lib/util/lifted/influx/httpd/handler_metrics.go:29]: 中文注释请改为英文

File: lib/util/lifted/influx/httpd/handler_metrics.go [hunk 2]
```diff
@@ -0,0 +1,246 @@
+package httpd
```
Comment (xiangyu5632) [inline lib/util/lifted/influx/httpd/handler_metrics.go:1]: 添加Apache License 头

**Comment by xuthus5** [comment]
Can I know the design plan of the metrics system?

File: lib/metrics/baseCollector.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+func NewBaseCollector(comIndexRegistry map[string]*Metric) *BaseCollector {
+	c := &BaseCollector{
+		AllModulesDesc: make(map[string]map[string]*prometheus.Desc),
+		IndexRegistry:  comIndexRegistry,
+	}
+
+	return c
+}
+
+func (c *BaseCollector) Collect(ch chan<- prometheus.Metric) {
```
Comment (xuthus5) [inline lib/metrics/baseCollector.go:56]: How does the `collector` react when the `/metrics` API is called? no specific logic here!

File: lib/util/lifted/influx/httpd/handler_metrics.go [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+	*metrics.BaseCollector
+	indexMap map[string]*metrics.ModuleIndex
+}
+
+func NewOpenGeminiCollector() *OpenGeminiCollector {
+	c := &OpenGeminiCollector{
+		BaseCollector: metrics.NewBaseCollector(metrics.ComIndexRegistry),
+		indexMap:      make(map[string]*metrics.ModuleIndex),
+	}
+	// 初始化指标
```
Comment (xuthus5) [inline lib/util/lifted/influx/httpd/handler_metrics.go]: Be more international and use English annotations

File: lib/metrics/index.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+		},
+		Labels: []string{"app", "hostname"},
+	},
+	"spdy": {
+		HelpMap: map[string]string{
+			"connTotal": "Total number of connections (internal)",
+		},
+		Labels: []string{"app", "hostname", "link", "remote_addr"},
+	},
+	"measurement_metric": {
```
Comment (xuthus5) [inline lib/metrics/index.go:103]: How to cancel the metrics when the `measurement` or `database` is deleted?

File: lib/metrics/index.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+		HelpMap: map[string]string{
+			"SeriesCount": "Number of timelines",
```
Comment (xuthus5) [inline lib/metrics/index.go:105]: more add and delete operation of mesurements cause metrics leak?

**Comment by xuthus5** [comment]
squash the commit into one?

**Comment by xiangyu5632** [comment]
> squash the commit into one?

yes

File: lib/metrics/baseCollector.go (hunk 3 above)
Comment (gueFDF) [reply lib/metrics/baseCollector.go:56]: This piece is an empty implementation, the specific logic is in handler_metrics.go

File: lib/metrics/index.go (hunk 5 above)
Comment (gueFDF) [reply lib/metrics/index.go:103]: If this database does not exist, it will be ignored.

**Comment by gueFDF** [comment]
> Can I know the design plan of the metrics system?

During initialization, the metric will be registered and initialized first. When ```/metric``` is called, ```getMetrics``` will be called to obtain the metric from the database, and then Collect will be triggered to write the metric through.

File: lib/util/lifted/influx/httpd/handler_metrics.go [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+			// with another result.  The series, on the other hand, still
+			// returns partial true if it was truncated or had more data to
+			// send in a future chunk.
+			r.Partial = false
+			break
+		}
+	}
+
+	resp := h.getStmtResult(stmtID2Result)
+	if resp.Results[0].Err != nil {
```
Comment (xuthus5) [inline lib/util/lifted/influx/httpd/handler_metrics.go:244]: Check the out-of-bounds array.

File: lib/util/lifted/influx/httpd/handler_metrics.go [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+		return nil, resp.Results[0].Err
+	}
+	metricsMap := make(map[string]interface{})
+
+	for i, key := range resp.Results[0].Series[0].Columns {
```
Comment (xuthus5) [inline lib/util/lifted/influx/httpd/handler_metrics.go:249]: ditto

File: lib/util/lifted/influx/httpd/handler_metrics.go [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+		metricsMap[key] = resp.Results[0].Series[0].Values[0][i]
+	}
+	labelValues := make(map[string]string)
```
Comment (xuthus5) [inline lib/util/lifted/influx/httpd/handler_metrics.go:252]: optimized: allocate the cap of the map in advance, when the Tags length is known,

File: lib/util/lifted/influx/httpd/handler_metrics.go [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+		if err != nil {
+			continue
+		}
+		openGeminiCollector.indexMap[moduleName] = moduleIndex
+	}
+
+	metricsHandler.ServeHTTP(w, r)
+}
+
+// serverProm
```
Comment (shoothzj) [inline lib/util/lifted/influx/httpd/handler_metrics.go:123]:
```suggestion
// getMetrics serverProm
```

File: lib/util/lifted/influx/httpd/handler_metrics.go [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+func getMetrics(h *Handler, r *http.Request, user meta2.User, tableName string, labelNames []string) (*metrics.ModuleIndex, error) {
+	// Retrieve the underlying ResponseWriter or initialize our own.
+	start := time.Now()
+
+	nodeID, _ := strconv.ParseUint(r.FormValue("node_id"), 10, 64)
```
Comment (shoothzj) [inline lib/util/lifted/influx/httpd/handler_metrics.go:128]: unhandle error

File: lib/metrics/index.go [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+var ComIndexRegistry = map[string]*Metric{
+	"httpd": {
+		HelpMap: map[string]string{
+			"fieldsWritten":       "The number of field columns written",
+			"writeReqBytes":       "Total amount of data written successfully (as Byte)",
+			"writeReqBytesIn":     "The total amount of write data (including write failures and write successes)",
+			"writeReq":            "The cumulative number of write requests received by the serverThe cumulative number of write requests received by the server",
+			"writeReqActive":      "The number of currently active write requests",
+			"writeReqDurationNs":  "Cumulative write data delay (NS)",
+			"pointsWrittenOK":     "Number of successful bars written",
```
Comment (shoothzj) [inline lib/metrics/index.go:28]: It's not a good name. pointsWrittenSuccess might be better, `OK` is more oral?

File: lib/metrics/index.go [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+			"queryReq":            "Number of client query requests received",
```
Comment (shoothzj) [inline lib/metrics/index.go:29]: queryReqCount?

File: lib/metrics/index.go [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+			"queryReqDurationNs":  "Cumulative query latency (NS)",
+			"queryReqActive":      "The number of currently active query requests",
+			"queryErrorStmtCount": "The number of query statements that failed the query",
```
Comment (shoothzj) [inline lib/metrics/index.go:32]: queryStmtErrorCount would be better?

File: lib/metrics/index.go [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+			"write400ErrReq":      "The cumulative number of errors caused by problems such as not finding the database, insufficient permissions, and so on",
+			"write500ErrReq":      "The cumulative number of write data errors within the server",
+			"queryStmtCount":      "Total number of query statements",
+		},
+		Labels: []string{"app", "hostname"},
+	},
+	"performance": {
+		HelpMap: map[string]string{
+			"WriteFieldsCount":        "The total number of indicator columns (Field) read on the ts-store node",
```
Comment (shoothzj) [inline lib/metrics/index.go:41]: Should we capitalize or lowercase?

File: lib/metrics/index.go [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+	"io": {
+		HelpMap: map[string]string{
+			"snapshotBytes": "Snapshot file size (cumulative, calculated in Byte)",
+		},
+		Labels: []string{"app", "hostname"},
+	},
+	"executor": {
+		HelpMap: map[string]string{
+			"exec_run_time_count": "The total number of executor executions of the query plan",
+			"exec_abort_count":    "The total number of query plan executor terminations",
```
Comment (shoothzj) [inline lib/metrics/index.go:62]: Should we snake style or camel style?

File: lib/util/lifted/influx/httpd/handler_metrics.go [hunk 17]
```diff
@@ -0,0 +1,263 @@
...
+	"strconv"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/openGemini/openGemini/lib/metrics"
+	"github.com/openGemini/openGemini/lib/statisticsPusher/statistics"
+	"github.com/openGemini/openGemini/lib/syscontrol"
+	"github.com/openGemini/openGemini/lib/util/lifted/influx/influxql"
+	meta2 "github.com/openGemini/openGemini/lib/util/lifted/influx/meta"
```
Comment (shoothzj) [inline lib/util/lifted/influx/httpd/handler_metrics.go:32]: is this alias needed?

**Review by shoothzj — CHANGES_REQUESTED** [review]
I saw there contains many comments, please solve it

**Comment by shoothzj** [comment]
I think a list of prometheus metrics definition should be reviewed first and carefully before implementation

File: lib/metrics/index.go (hunk 12 above)
Comment (xiangyu5632) [reply lib/metrics/index.go:28]: These are the names of the kernel monitoring metrics collected by openGemini. They already exist in the database and have not been newly added by this function.
To standardize the names of these metrics, we need to modify the code of the kernel collection metrics and the matching Grafana monitoring panel. This issue is not related to the functions developed this time. But we may submit another PR for modification.

File: lib/metrics/index.go (hunk 13 above)
Comment (xiangyu5632) [reply lib/metrics/index.go:29]: as the same as above

File: lib/metrics/index.go (hunk 14 above)
Comment (xiangyu5632) [reply lib/metrics/index.go:32]: as the same as above

File: lib/util/lifted/influx/httpd/handler_metrics.go [hunk 18]
```diff
@@ -0,0 +1,263 @@
...
+		defer func() {
+			d := time.Now().Sub(start)
+			if d.Nanoseconds() > time.Second.Nanoseconds()*10 {
+				qDuration.AddDuration("TotalDuration", d.Nanoseconds())
+				statistics.AppendSqlQueryDuration(qDuration)
+				h.Logger.Info("slow query", zap.Duration("duration", d), zap.String("db", qDuration.DB),
+					zap.String("query", qDuration.Query))
+			}
+		}()
+	}
```
Comment (shuangquansq) [inline lib/util/lifted/influx/httpd/handler_metrics.go:159]: the db name must be "_internal", so this code can be deleted.

File: lib/util/lifted/influx/httpd/handler_metrics.go [hunk 19]
```diff
@@ -0,0 +1,263 @@
...
+
+}
+
+func (h *Handler) serveMetrics(w http.ResponseWriter, r *http.Request, user meta2.User) {
+	for moduleName, index := range openGeminiCollector.IndexRegistry {
+		moduleIndex, err := getMetrics(h, r, user, moduleName, index.Labels)
+		if err != nil {
+			continue
+		}
+		openGeminiCollector.indexMap[moduleName] = moduleIndex
```
Comment (shuangquansq) [inline lib/util/lifted/influx/httpd/handler_metrics.go:117]: For the metrics, add the time info, like this : 
writeRowsCount{app="ts-single",hostname="127.0.0.1:8401"} 23 1713163245665
I debug your program, no time info:
writeRowsCount{app="ts-single",hostname="127.0.0.1:8401"} 23

Comment (gueFDF) [reply lib/util/lifted/influx/httpd/handler_metrics.go:117]: I'm not sure if prom can consume data in this format.

File: lib/util/lifted/influx/httpd/handler_metrics.go [hunk 20]
```diff
@@ -0,0 +1,263 @@
...
+
+	qr := strings.NewReader(sql)
+	q, err, _ = h.getSqlQuery(r, qr)
+	if err != nil {
+		return nil, err
+	}
+
+	db := h.SQLConfig.Monitor.StoreDatabase
+	var qDuration *statistics.SQLSlowQueryStatistics
+	if !isInternalDatabase(db) {
```
Comment (shuangquansq) [inline lib/util/lifted/influx/httpd/handler_metrics.go:148]:
```suggestion
		if !isInternalDatabase(db) {
		qDuration = statistics.NewSqlSlowQueryStatistics(db)
		defer func() {
			d := time.Now().Sub(start)
			if d.Nanoseconds() > time.Second.Nanoseconds()*10 {
				qDuration.AddDuration("TotalDuration", d.Nanoseconds())
				statistics.AppendSqlQueryDuration(qDuration)
				h.Logger.Info("slow query", zap.Duration("duration", d), zap.String("db", qDuration.DB),
					zap.String("query", qDuration.Query))
			}
		}()
	}
```
The db name must be "_internal", this code can be deleted. The code record slowly sql for business db.

File: lib/util/lifted/influx/httpd/handler_metrics.go [hunk 21]
```diff
@@ -0,0 +1,263 @@
...
+		openGeminiCollector.indexMap[moduleName] = moduleIndex
+	}
+
+	metricsHandler.ServeHTTP(w, r)
+}
+
+// serverProm
+func getMetrics(h *Handler, r *http.Request, user meta2.User, tableName string, labelNames []string) (*metrics.ModuleIndex, error) {
+	// Retrieve the underlying ResponseWriter or initialize our own.
+	start := time.Now()
```
Comment (shuangquansq) [inline lib/util/lifted/influx/httpd/handler_metrics.go:126]: No need for "_internal" db.

Comment (shuangquansq) [inline lib/util/lifted/influx/httpd/handler_metrics.go:126]:
```suggestion
	start := time.Now()
```
No need for "_internal" db.

File: lib/util/lifted/influx/httpd/handler_metrics.go (hunk 19 above)
Comment (shuangquansq) [inline lib/util/lifted/influx/httpd/handler_metrics.go:117]:
```suggestion
		openGeminiCollector.indexMap[moduleName] = moduleIndex
```
For the metrics, please add the time info, like this :
writeRowsCount{app="ts-single",hostname="127.0.0.1:8401"} 23 1713163245665
I debug your program, no time info for metric:
writeRowsCount{app="ts-single",hostname="127.0.0.1:8401"} 23
we need time map to metric for analyze.

Comment (shuangquansq) [inline lib/util/lifted/influx/httpd/handler_metrics.go:117]: For the metrics, please add the time info, like this :
writeRowsCount{app="ts-single",hostname="127.0.0.1:8401"} 23 1713163245665
I debug your program, no time info for metric:
writeRowsCount{app="ts-single",hostname="127.0.0.1:8401"} 23
we need time map to metric for analyze.

File: Makefile.common [hunk 22]
```diff
@@ -31,7 +31,7 @@ PACKAGES_OPEN_GEMINI_TESTS ?= $$($(PACKAGE_LIST_OPEN_GEMINI_TESTS))
 
 COPYRIGHT_EXCEPT  := "lifted|tests|lib/netstorage/data/data.pb.go|lib/statisticsPusher/statistics/handler_statistics.go|app/ts-meta/meta/snapshot.go|engine/index/tsi/tag_filters.go|engine/index/tsi/tag_filter_test.go|engine/index/mergeindex/item.go|lib/config/openGemini_dir.go"
 COPYRIGHT_GOFILE  := $$(find . -name '*.go' | grep -vE $(COPYRIGHT_EXCEPT))
-COPYRIGHT_HEADER  := "Copyright 2022|2023|2024 Huawei Cloud Computing Technologies Co., Ltd."
+COPYRIGHT_HEADER  := "Copyright 2022|2023|2024 Huawei Cloud Computing Technologies Co., Ltd.|Copyright right 2024 openGemini author."
```
Comment (xiangyu5632) [inline Makefile.common:34]: > Copyright right 2024 openGemini author. 
多了一个right

**Comment by xuthus5** [comment]
please squash your commit. @gueFDF

File: lib/util/lifted/influx/httpd/handler_metrics.go (hunk 4 above)
Comment (gueFDF) [reply lib/util/lifted/influx/httpd/handler_metrics.go]: Okay, I've solved it.

File: lib/util/lifted/influx/httpd/handler_metrics.go (hunk 7 above)
Comment (gueFDF) [reply lib/util/lifted/influx/httpd/handler_metrics.go:234]: Okay, I've solved it. I used for-range o avoid.

File: lib/util/lifted/influx/httpd/handler_metrics.go (hunk 8 above)
Comment (gueFDF) [reply lib/util/lifted/influx/httpd/handler_metrics.go:239]: Okay, I've solved it.

File: lib/util/lifted/influx/httpd/handler_metrics.go (hunk 10 above)
Comment (gueFDF) [reply lib/util/lifted/influx/httpd/handler_metrics.go]: Okay, I've solved it.

File: lib/metrics/index.go (hunk 15 above)
Comment (gueFDF) [reply lib/metrics/index.go]: as the same as above

File: lib/util/lifted/influx/httpd/handler_metrics.go (hunk 17 above)
Comment (gueFDF) [reply lib/util/lifted/influx/httpd/handler_metrics.go]: as the same as above

File: lib/util/lifted/influx/httpd/handler_metrics.go (hunk 9 above)
Comment (gueFDF) [reply lib/util/lifted/influx/httpd/handler_metrics.go:239]: OK, this code logic has been modified.

File: lib/util/lifted/influx/httpd/handler_metrics.go (hunk 11 above)
Comment (gueFDF) [reply lib/util/lifted/influx/httpd/handler_metrics.go:129]: This block is checked in the following execution function, which is consistent with the rest of the code.

File: lib/util/lifted/influx/httpd/handler_metrics.go (hunk 20 above)
Comment (gueFDF) [reply lib/util/lifted/influx/httpd/handler_metrics.go]: I have removed.

File: lib/util/lifted/influx/httpd/handler_metrics.go (hunk 19 above)
Comment (gueFDF) [reply lib/util/lifted/influx/httpd/handler_metrics.go:121]: This piece of logic has been added.

File: lib/util/lifted/influx/httpd/handler_metrics.go [hunk 23]
```diff
@@ -0,0 +1,259 @@
...
+	var err error
+	sql := fmt.Sprintf("select * from %s", tableName)
+	for i, labelName := range labelNames {
+		if i == 0 {
+			sql = fmt.Sprintf("%s group by %s", sql, labelName)
+		} else {
+			sql = fmt.Sprintf("%s,%s", sql, labelName)
+		}
+	}
+	sql = fmt.Sprintf("%s order by time DESC limit 1", sql)
```
Comment (xiangyu5632) [inline lib/util/lifted/influx/httpd/handler_metrics.go:140]: this query may return one of hostname data，we need return all hostname's new data, so we can change the query to 
```
select top(time, 1), reqActive, reqDurationNs, xxx from tableName group by hostname where time >= now() - 1m
```

Comment (gueFDF) [reply lib/util/lifted/influx/httpd/handler_metrics.go]: Okay, I've solved it.

**Comment by xiangyu5632** [comment]
@gueFDF do you submit all the modifications? should I merge?

**Comment by gueFDF** [comment]
> @gueFDF do you submit all the modifications? should I merge?

Yes, I've committed all my changes

**Review by xiangyu5632 — APPROVED** [review]

**Review by shoothzj — APPROVED** [review]

**Pull request merged**